File: GateSyntax/csharp/GateSyntaxIntegrador.cs
// GateSyntaxIntegrador.cs — C# / WPF
// Domain-agnostic GateSyntax integrador.
//
// Drop into any C# project that references GateSyntax-CSharp.
// Reflects on [GsExpose]-annotated members, auto-generates .ui declarations,
// and runs a live sync loop so the UI always mirrors the host program's state.
//
// Usage:
//   var ig = new GateSyntaxIntegrador(this);   // reflect on current object
//   ig.Run();                                   // opens the live GateSyntax window
//
// Or fluent registration:
//   GateSyntaxIntegrador.Create()
//       .Bind("speed",  () => _speed,  v => _speed = v,  min: 0, max: 200)
//       .Action("Reset", () => _speed = 0)
//       .Run();

using System;
using System.Collections.Generic;
using System.Reflection;
using System.Threading;
using System.Windows.Threading;
using GateSyntax.Runtime;          // GateSyntax-CSharp runtime

namespace GateSyntax
{
    // ── Attribute ─────────────────────────────────────────────────────────────

    [AttributeUsage(AttributeTargets.Property | AttributeTargets.Field | AttributeTargets.Method)]
    public sealed class GsExposeAttribute : Attribute
    {
        public string? Label { get; init; }
        public double  Min   { get; init; } = 0;
        public double  Max   { get; init; } = 100;
        public string? Group { get; init; }
    }

    // ── Binding descriptor ────────────────────────────────────────────────────

    internal sealed class GsBinding
    {
        public string          Name    { get; init; } = "";
        public string          Label   { get; init; } = "";
        public Func<object?>?  Getter  { get; init; }
        public Action<object>? Setter  { get; init; }
        public Action?         Action  { get; init; }
        public double          Min     { get; init; } = 0;
        public double          Max     { get; init; } = 100;
        public Type?           Type    { get; init; }

        public bool IsAction  => Action  != null;
        public bool IsNumeric => Type == typeof(double) || Type == typeof(float)
                                  || Type == typeof(int) || Type == typeof(long);
        public bool IsBool    => Type == typeof(bool);
    }

    // ── Integrador ────────────────────────────────────────────────────────────

    public sealed class GateSyntaxIntegrador
    {
        private readonly List<GsBinding>     _bindings = new();
        private          StateStore?         _store;
        private          bool                _running;
        private readonly double              _pollHz;

        public GateSyntaxIntegrador(double pollHz = 30) { _pollHz = pollHz; }

        // ── Fluent registration ───────────────────────────────────────────────

        public GateSyntaxIntegrador Bind(string name, Func<object?> getter,
            Action<object>? setter = null, double min = 0, double max = 100,
            string? label = null, Type? type = null)
        {
            _bindings.Add(new GsBinding
            {
                Name = name, Label = label ?? name,
                Getter = getter, Setter = setter,
                Min = min, Max = max,
                Type = type ?? getter()?.GetType() ?? typeof(string),
            });
            return this;
        }

        public GateSyntaxIntegrador Action(string name, Action fn, string? label = null)
        {
            _bindings.Add(new GsBinding { Name = name, Label = label ?? name, Action = fn });
            return this;
        }

        // ── Reflection discovery ──────────────────────────────────────────────

        public static GateSyntaxIntegrador FromObject(object host, double pollHz = 30)
        {
            var ig = new GateSyntaxIntegrador(pollHz);
            var type = host.GetType();

            foreach (var prop in type.GetProperties(BindingFlags.Public | BindingFlags.Instance))
            {
                var attr = prop.GetCustomAttribute<GsExposeAttribute>();
                if (attr == null) continue;
                ig.Bind(
                    name:   prop.Name,
                    getter: () => prop.GetValue(host),
                    setter: prop.CanWrite ? v => prop.SetValue(host, Convert.ChangeType(v, prop.PropertyType)) : null,
                    min:    attr.Min, max: attr.Max,
                    label:  attr.Label ?? prop.Name,
                    type:   prop.PropertyType
                );
            }

            foreach (var method in type.GetMethods(BindingFlags.Public | BindingFlags.Instance))
            {
                var attr = method.GetCustomAttribute<GsExposeAttribute>();
                if (attr == null || method.GetParameters().Length > 0) continue;
                ig.Action(method.Name, () => method.Invoke(host, null), attr.Label ?? method.Name);
            }

            return ig;
        }

        // ── UI generation ─────────────────────────────────────────────────────

        private string GenerateUi()
        {
            var sb = new System.Text.StringBuilder();
            sb.AppendLine("WINDOW Root :: TITLE \"GateSyntax Integrador\"");
            sb.AppendLine("SCROLL MainScroll :: IN [Root]");
            sb.AppendLine("COL    MainCol    :: IN [MainScroll]");

            foreach (var b in _bindings)
            {
                string var_ = $"GS_{b.Name.ToUpper()}";

                if (b.IsAction)
                {
                    sb.AppendLine($"BUTTON {b.Name}Btn :: IN [MainCol] :: LABEL \"▶  {b.Label}\" :: ON CLICK /{var_}_CALL :: \"CALL_{b.Name.ToUpper()}\"\\");
                }
                else if (b.IsNumeric)
                {
                    double init = Convert.ToDouble(b.Getter?.Invoke() ?? b.Min);
                    sb.AppendLine($"/{var_} :: {init}\\");
                    sb.AppendLine($"LABEL  {b.Name}Lbl :: IN [MainCol] :: TEXT \"{b.Label}:  \" + [{var_}]");
                    sb.AppendLine($"SLIDER {b.Name}Sl  :: IN [MainCol] :: MIN {b.Min} :: MAX {b.Max} :: VALUE [{var_}] :: ON CHANGE /{var_} :: [{b.Name}Sl]\\");
                    sb.AppendLine($"RULE   {b.Name}Sep :: IN [MainCol]");
                }
                else if (b.IsBool)
                {
                    bool init = b.Getter?.Invoke() is true;
                    sb.AppendLine($"/{var_} :: {(init ? "TRUE" : "FALSE")}\\");
                    sb.AppendLine($"TOGGLE {b.Name}Tog :: IN [MainCol] :: LABEL \"{b.Label}\" :: VALUE [{var_}] :: ON CHANGE /{var_} :: [{b.Name}Tog]\\");
                }
                else
                {
                    string init = b.Getter?.Invoke()?.ToString() ?? "";
                    sb.AppendLine($"/{var_} :: \"{init}\"\\");
                    sb.AppendLine($"LABEL {b.Name}Lbl :: IN [MainCol] :: TEXT \"{b.Label}\"");
                    sb.AppendLine($"INPUT {b.Name}In  :: IN [MainCol] :: HINT \"Enter {b.Label}…\" :: ON CHANGE /{var_} :: [{b.Name}In]\\");
                }
            }

            return sb.ToString();
        }

        // ── Live sync loop ────────────────────────────────────────────────────

        private void LiveLoop()
        {
            double interval = 1000.0 / _pollHz;
            while (_running)
            {
                foreach (var b in _bindings)
                {
                    if (b.Getter == null) continue;
                    try
                    {
                        var current = b.Getter();
                        _store?.Set($"GS_{b.Name.ToUpper()}", current ?? "");
                    }
                    catch { /* host threw; skip */ }
                }
                Thread.Sleep((int)interval);
            }
        }

        // ── Entry point ───────────────────────────────────────────────────────

        public void Run()
        {
            var uiContent = GenerateUi();
            var (app, store) = new GateSyntaxBuilder()
                .AddContent(uiContent)
                .BuildWithState();

            _store = store;

            // Wire setters: UI change → host update
            foreach (var b in _bindings)
            {
                if (b.Setter == null) continue;
                var varName = $"GS_{b.Name.ToUpper()}";
                var setter  = b.Setter;
                store.Subscribe(varName, v => setter(v));
            }

            // Register action handlers
            foreach (var b in _bindings)
            {
                if (!b.IsAction) continue;
                var actionKey = $"CALL_{b.Name.ToUpper()}";
                var fn = b.Action!;
                app.RegisterAction(actionKey, fn);
            }

            // Start live poll thread
            _running = true;
            var thread = new Thread(LiveLoop) { IsBackground = true };
            thread.Start();

            app.Run();
            _running = false;
        }

        // ── Static factories ──────────────────────────────────────────────────

        public static GateSyntaxIntegrador Create(double pollHz = 30) => new(pollHz);
        public static void RunFor(object host) => FromObject(host).Run();
    }
}

File: GateSyntax/java/src/main/java/com/gatesyntax/integrador/GateSyntaxIntegrador.java
// GateSyntaxIntegrador.java — Java / JavaFX
// Domain-agnostic GateSyntax integrador.
//
// Drop into any Java project that has GateSyntaxJava on the classpath.
// Reflects on @GsExpose-annotated members, auto-generates .ui declarations,
// and runs a live sync loop so the JavaFX UI always mirrors the host program.
//
// Usage:
//   GateSyntaxIntegrador.runFor(this);         // reflect + launch
//
// Or fluent:
//   GateSyntaxIntegrador.create()
//       .bind("speed",  () -> speed,  v -> speed = (double)v, 0, 200)
//       .action("Reset", () -> speed = 0)
//       .run();

package com.gatesyntax.integrador;

import com.gatesyntax.GateSyntaxApp;
import com.gatesyntax.GateSyntaxBuilder;
import com.gatesyntax.runtime.*;
import javafx.application.Application;
import javafx.application.Platform;

import java.lang.reflect.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// ── Integrador ────────────────────────────────────────────────────────────────

public final class GateSyntaxIntegrador {

    // ── Binding descriptor ────────────────────────────────────────────────────

    private record Binding(
        String            name,
        String            label,
        Supplier<Object>  getter,
        Consumer<Object>  setter,
        Runnable          action,
        double            min,
        double            max,
        Class<?>          type
    ) {
        boolean isAction()  { return action != null; }
        boolean isNumeric() { return type == double.class || type == float.class
                                  || type == int.class    || type == long.class
                                  || Number.class.isAssignableFrom(type); }
        boolean isBool()    { return type == boolean.class || type == Boolean.class; }
    }

    // ── State ─────────────────────────────────────────────────────────────────

    private final List<Binding>          bindings  = new ArrayList<>();
    private       StateStore             store;
    private       ScheduledExecutorService ticker;
    private final double                 pollHz;

    public GateSyntaxIntegrador(double pollHz) { this.pollHz = pollHz; }

    // ── Fluent API ────────────────────────────────────────────────────────────

    public GateSyntaxIntegrador bind(String name, Supplier<Object> getter,
            Consumer<Object> setter, double min, double max) {
        return bind(name, name, getter, setter, min, max,
                    getter.get() != null ? getter.get().getClass() : String.class);
    }

    public GateSyntaxIntegrador bind(String name, String label,
            Supplier<Object> getter, Consumer<Object> setter,
            double min, double max, Class<?> type) {
        bindings.add(new Binding(name, label, getter, setter, null, min, max, type));
        return this;
    }

    public GateSyntaxIntegrador action(String name, Runnable fn) {
        return action(name, name, fn);
    }

    public GateSyntaxIntegrador action(String name, String label, Runnable fn) {
        bindings.add(new Binding(name, label, null, null, fn, 0, 100, Void.class));
        return this;
    }

    // ── Reflection discovery ──────────────────────────────────────────────────

    public static GateSyntaxIntegrador fromObject(Object host) {
        var ig = create();

        for (Field f : host.getClass().getDeclaredFields()) {
            var ann = f.getAnnotation(GsExpose.class);
            if (ann == null) continue;
            f.setAccessible(true);
            String label = ann.label().isEmpty() ? f.getName() : ann.label();
            ig.bind(f.getName(), label,
                () -> { try { return f.get(host); } catch (Exception e) { return null; } },
                v  -> { try { f.set(host, castTo(v, f.getType())); } catch (Exception ignored) {} },
                ann.min(), ann.max(), f.getType());
        }

        for (Method m : host.getClass().getDeclaredMethods()) {
            var ann = m.getAnnotation(GsExpose.class);
            if (ann == null || m.getParameterCount() > 0) continue;
            m.setAccessible(true);
            String label = ann.label().isEmpty() ? m.getName() : ann.label();
            ig.action(m.getName(), label, () -> {
                try { m.invoke(host); } catch (Exception ignored) {}
            });
        }

        return ig;
    }

    // ── .ui generation ────────────────────────────────────────────────────────

    private String generateUi() {
        var sb = new StringBuilder();
        sb.append("WINDOW Root :: TITLE \"GateSyntax Integrador\"\n");
        sb.append("SCROLL MainScroll :: IN [Root]\n");
        sb.append("COL    MainCol    :: IN [MainScroll]\n");

        for (var b : bindings) {
            String var_ = "GS_" + b.name().toUpperCase();

            if (b.isAction()) {
                sb.append("BUTTON ").append(b.name()).append("Btn :: IN [MainCol]")
                  .append(" :: LABEL \"▶  ").append(b.label()).append("\"")
                  .append(" :: ON CLICK /").append(var_).append("_CALL")
                  .append(" :: \"CALL_").append(b.name().toUpperCase()).append("\"\\\n");

            } else if (b.isNumeric()) {
                double init = b.getter() != null
                    ? ((Number) b.getter().get()).doubleValue() : b.min();
                sb.append("/").append(var_).append(" :: ").append(init).append("\\\n");
                sb.append("LABEL ").append(b.name()).append("Lbl :: IN [MainCol]")
                  .append(" :: TEXT \"").append(b.label()).append(":  \" + [").append(var_).append("]\n");
                sb.append("SLIDER ").append(b.name()).append("Sl :: IN [MainCol]")
                  .append(" :: MIN ").append(b.min()).append(" :: MAX ").append(b.max())
                  .append(" :: VALUE [").append(var_).append("]")
                  .append(" :: ON CHANGE /").append(var_).append(" :: [").append(b.name()).append("Sl]\\\n");
                sb.append("RULE ").append(b.name()).append("Sep :: IN [MainCol]\n");

            } else if (b.isBool()) {
                boolean init = b.getter() != null && Boolean.TRUE.equals(b.getter().get());
                sb.append("/").append(var_).append(" :: ").append(init ? "TRUE" : "FALSE").append("\\\n");
                sb.append("TOGGLE ").append(b.name()).append("Tog :: IN [MainCol]")
                  .append(" :: LABEL \"").append(b.label()).append("\"")
                  .append(" :: VALUE [").append(var_).append("]")
                  .append(" :: ON CHANGE /").append(var_).append(" :: [").append(b.name()).append("Tog]\\\n");

            } else {
                String init = b.getter() != null ? String.valueOf(b.getter().get()) : "";
                sb.append("/").append(var_).append(" :: \"").append(init).append("\"\\\n");
                sb.append("LABEL ").append(b.name()).append("Lbl :: IN [MainCol]")
                  .append(" :: TEXT \"").append(b.label()).append("\"\n");
                sb.append("INPUT ").append(b.name()).append("In :: IN [MainCol]")
                  .append(" :: HINT \"Enter ").append(b.label()).append("…\"")
                  .append(" :: ON CHANGE /").append(var_).append(" :: [").append(b.name()).append("In]\\\n");
            }
        }
        return sb.toString();
    }

    // ── Live poll loop ────────────────────────────────────────────────────────

    private void startLiveLoop() {
        long periodMs = (long) (1000.0 / pollHz);
        ticker = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "gs-integrador-poll");
            t.setDaemon(true);
            return t;
        });
        ticker.scheduleAtFixedRate(() -> {
            for (var b : bindings) {
                if (b.getter() == null) continue;
                try {
                    Object val = b.getter().get();
                    Platform.runLater(() -> store.set("GS_" + b.name().toUpperCase(), val));
                } catch (Exception ignored) {}
            }
        }, 0, periodMs, TimeUnit.MILLISECONDS);
    }

    // ── Entry point ───────────────────────────────────────────────────────────

    public void run() {
        String uiContent = generateUi();
        var builder = new GateSyntaxBuilder().addContent(uiContent);
        var runtime  = builder.buildRuntime();
        store = runtime.getStore();

        // UI → host: subscribe to state changes
        for (var b : bindings) {
            if (b.setter() == null) continue;
            String varName = "GS_" + b.name().toUpperCase();
            Consumer<Object> setter = b.setter();
            store.subscribe(varName, v -> {
                try { setter.accept(castTo(v, b.type())); } catch (Exception ignored) {}
            });
        }

        // Register action handlers
        for (var b : bindings) {
            if (!b.isAction()) continue;
            String key = "CALL_" + b.name().toUpperCase();
            Runnable fn = b.action();
            runtime.registerAction(key, fn);
        }

        // Start host → UI live poll
        startLiveLoop();

        // Launch JavaFX
        GateSyntaxApp.pendingBuilder = builder;
        Application.launch(GateSyntaxApp.class);

        ticker.shutdownNow();
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private static Object castTo(Object v, Class<?> type) {
        if (v == null) return null;
        if (type == double.class || type == Double.class)  return ExpressionEvaluator.toDouble(v);
        if (type == float.class  || type == Float.class)   return (float) ExpressionEvaluator.toDouble(v);
        if (type == int.class    || type == Integer.class) return (int) ExpressionEvaluator.toDouble(v);
        if (type == long.class   || type == Long.class)    return (long) ExpressionEvaluator.toDouble(v);
        if (type == boolean.class|| type == Boolean.class) return ExpressionEvaluator.toBool(v);
        return ExpressionEvaluator.toStr(v);
    }

    private Supplier<Object> getter() { return null; } // unused placeholder

    // ── Static factories ──────────────────────────────────────────────────────

    public static GateSyntaxIntegrador create()               { return new GateSyntaxIntegrador(30); }
    public static GateSyntaxIntegrador create(double hz)      { return new GateSyntaxIntegrador(hz); }
    public static void                 runFor(Object host)    { fromObject(host).run(); }
}
